bayes_kit/rhat.py

```diff
@@ -1,11 +1,88 @@
 import numpy as np
 from numpy.typing import NDArray, ArrayLike
+import scipy as sp
 
 FloatType = np.float64
 VectorType = NDArray[FloatType]
 SeqType = ArrayLike
 
 
+def split_chains(chains: list[SeqType]) -> list[SeqType]:
+    """
+    Return a list of the input chains split in half.  The result will be
+    a list twice as long as the input.  For example,
+    ```
+    >>> split_chains([[1, 2, 3, 4], [5, 6, 7]])
+    [[1, 2], [3, 4], [5, 6], [7]]
+    ```
+
+    Parameters:
+    chains: List of univariate Markov chains.
+
+    Returns:
+    List of input chains split in half.
+    """
+    return [list(arr) for chain in chains for arr in np.array_split(chain, 2)]
+
+
+def rank_chains(chains: list[SeqType]) -> list[SeqType]:
+    """
+    Returns a copy of the included Markov chains transformed with
+    ranks normalized to [0, 1] and an offset inverse CDF. The ranks
+    are ascending and start with 1 for the smallest value.
+
+    Parameters:
+    chains: list of univariate Markov chains
+
+    Returns:
+    List of chains with values replaced by transformed ranks.
+    """
+    if len(chains) == 0:
+        return chains
+    flattened = np.concatenate(chains)
+    ranks = flattened.argsort().argsort() + 1
```

Comment:
Tests fail with only a single .argsort().  This is a bit worrying in that I asked ChatGPT to write this function and I wrote exhaustive tests, but I clearly don't understand it well enough. 

 Here's an example to show why we need two applications of `.argsort()`:

```python
>>> chains = [[4.2, 5.7], [7.2, 6.1], [-12.9, 107]]

>>> flattened = np.concatenate(chains)

>>> flattened
array([  4.2,   5.7,   7.2,   6.1, -12.9, 107. ])

>>> flattened.argsort()
array([4, 0, 1, 3, 2, 5])

>>> flattened.argsort().argsort()
array([1, 2, 4, 3, 0, 5])
```

To help, the doc notes that if you have an array and argsort and then apply, you get something in sorted order.

```python
>>> flattened[flattened.argsort()]
array([-12.9,   4.2,   5.7,   6.1,   7.2, 107. ])
```

So the first `.argsort()` makes sense.  But then I'm not sure why we don't get a fixed point after one application or why that works with the rest of the code.

Here's the doc:  https://numpy.org/doc/stable/reference/generated/numpy.argsort.html

The `+1` is because we're indexing from 1 in the transform.

Reply:
I updated my comment to explain why it works, but our comments crossed paths.  @roualdes sorted me out and I understand what it's doing now.  Apparently this is a standard trick.  I'm a bit reluctant to include stackoverflow links in my doc.